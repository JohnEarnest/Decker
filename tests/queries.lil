
# even more dedicated query tests

# multiple orderby:
t1:insert name id with
 "Banana" 22
 "Durian" 11
 "Cherry" 33
 "Apple"  55
 "Apple"  44
 "Banana" 88
end
show[select name id orderby name asc orderby id desc from t1]

# handling singles, empties:
show[extract first value by value from 11,22,11]
show[extract first value by value from 11]
show[extract first value by value from ()]

# tolerate undersized column expression results:
show[extract id by      ()     from t1]
show[extract id where   ()     from t1]
show[extract id orderby () asc from t1]
show[extract id by      7      from t1]
show[extract id where   7      from t1]
show[extract id orderby 7  asc from t1]
show["evens" extract id where 1,0 from t1]
show["odds"  extract id where 0,1 from t1]

# misc update scenarios:
show[update name:"Best %s" format name where gindex=0 orderby id desc by name from t1]
show[update name:"Best %s" format name where gindex=0 where 1<count gindex orderby id desc by name from t1]
show[update best:1                     where gindex=0 where 1<count gindex orderby id desc by name from t1]

# the perl weekly challenge 233, part 2: frequency sort
# (disambiguates identical freqs by sorting by value)
on f x do
 extract v orderby c asc orderby v desc from
  select v:value c:count value by value from x
end
show[f[(1,1,2,2,2,3         )]]
show[f[(2,3,1,3,2           )]]
show[f[(-1,1,-6,4,5,-6,1,4,1)]]

# examples from the lil reference manual:
people:insert name age job with
 "Alice"  25 "Developer"
 "Sam"    28 "Sales"
 "Thomas" 40 "Developer"
 "Sara"   34 "Developer"
 "Walter" 43 "Accounting"
end
jobs:insert job salary with
 "Sales"      85000
 "Developer"  75000
 "Accounting" 60000
 "Facilities" 50000
end
guests:insert name with "Alice" "Joan" "Oscar" "Thomas" end
show[select name where ("S%m" parse name) from people]
show[select name index orderby name asc from people]
show[select name job by job orderby name asc from people]
show[select name job orderby (job join name) asc from people]
show[select employed:(count name) job:(first job) by job from people]
show[extract v:value i:index orderby value asc  from extract first job by job from people]
show[extract value           orderby index desc from extract first job by job from people]
show[extract list index by value from "ABBAAC"]
show[extract list value by floor index/3 from "ABCDEFGHI"]
show[extract first value by value from "ABBAAC"]
show[extract first value by value from "AA"]
show[extract first value by value from ""]
show[select a:name b:name_ where name < name_ from guests cross guests]
denormal:select "with \"escapes":index "count":value from "ABC"
show[select where column["with \"escapes"]>0 from denormal]
show[first extract column from denormal]

# group, then regroup
t2:table("a","b","c")dict(
 (list "" split "ABABABABABABAB"),
 (list "" split "12341234123412"),
 (list "" split "abcdefghijklmn")
)
show[select (first b) list group by b from t2]
show[select where gindex=0 by b from t2]
show[select where gindex=0 by a where gindex=0 by b from t2]

# multiple where clauses for progressive narrowing
tested:()
on track x do tested:tested,count x x end
show[select where track[b]="1" where track[a]="A" from t2]
show[tested]

# gindex is effectively "reset" at each clause:
show[select a b gindex orderby a asc from t2]
show[select a b gindex where a="A" from t2]

# group with no rows:
show[select by a from 0 take t2]
show[select by a where a="Z" from t2]

# insert edge-cases:
show[insert with into 0]                           # no column names, empty
show[insert with 42 end]                           # no column names, single
show[insert a b c with 11 22 33 44 end]            # ragged value count -> pad with nil
show[insert a "quoted name" with 55 66 77 88 end]  # quoted column names
show[insert a a with "one" 11 "two" 22 end]        # repeat column names: take the last one

show[insert __proto__ with 11 22 end] # cursed column names
