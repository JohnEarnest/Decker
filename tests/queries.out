+----------+----+
| name     | id |
+----------+----+
| "Apple"  | 55 |
| "Apple"  | 44 |
| "Banana" | 88 |
| "Banana" | 22 |
| "Cherry" | 33 |
| "Durian" | 11 |
+----------+----+
(11,22)
(11)
()
(22,11,33,55,44,88)
()
(22,11,33,55,44,88)
(22,11,33,55,44,88)
(22,11,33,55,44,88)
(22,11,33,55,44,88)
"evens" (22,33,44)
"odds" (11,55,88)
+---------------+----+
| name          | id |
+---------------+----+
| "Banana"      | 22 |
| "Best Durian" | 11 |
| "Best Cherry" | 33 |
| "Best Apple"  | 55 |
| "Apple"       | 44 |
| "Best Banana" | 88 |
+---------------+----+
+---------------+----+
| name          | id |
+---------------+----+
| "Banana"      | 22 |
| "Durian"      | 11 |
| "Cherry"      | 33 |
| "Best Apple"  | 55 |
| "Apple"       | 44 |
| "Best Banana" | 88 |
+---------------+----+
+----------+----+------+
| name     | id | best |
+----------+----+------+
| "Banana" | 22 | nil  |
| "Durian" | 11 | nil  |
| "Cherry" | 33 | nil  |
| "Apple"  | 55 | 1    |
| "Apple"  | 44 | nil  |
| "Banana" | 88 | 1    |
+----------+----+------+
(3,1,1,2,2,2)
(1,3,3,2,2)
(5,-1,4,4,-6,-6,1,1,1)
+--------+
| name   |
+--------+
| "Sam"  |
| "Sara" |
+--------+
+----------+-------+
| name     | index |
+----------+-------+
| "Alice"  | 0     |
| "Sam"    | 1     |
| "Sara"   | 3     |
| "Thomas" | 2     |
| "Walter" | 4     |
+----------+-------+
+----------+--------------+
| name     | job          |
+----------+--------------+
| "Alice"  | "Developer"  |
| "Sara"   | "Developer"  |
| "Thomas" | "Developer"  |
| "Sam"    | "Sales"      |
| "Walter" | "Accounting" |
+----------+--------------+
+----------+--------------+
| name     | job          |
+----------+--------------+
| "Walter" | "Accounting" |
| "Alice"  | "Developer"  |
| "Sara"   | "Developer"  |
| "Thomas" | "Developer"  |
| "Sam"    | "Sales"      |
+----------+--------------+
+----------+--------------+
| employed | job          |
+----------+--------------+
| 3        | "Developer"  |
| 1        | "Sales"      |
| 1        | "Accounting" |
+----------+--------------+
{"v":("Accounting","Developer","Sales"),"i":(2,0,1)}
("Accounting","Sales","Developer")
((0,3,4),(1,2),(5))
(("A","B","C"),("D","E","F"),("G","H","I"))
("A","B","C")
("A")
()
+---------+----------+
| a       | b        |
+---------+----------+
| "Alice" | "Joan"   |
| "Alice" | "Oscar"  |
| "Joan"  | "Oscar"  |
| "Alice" | "Thomas" |
| "Joan"  | "Thomas" |
| "Oscar" | "Thomas" |
+---------+----------+
+---------------+-------+
| with "escapes | count |
+---------------+-------+
| 1             | "B"   |
| 2             | "C"   |
+---------------+-------+
+---------------+-------+-------+--------+-------+
| with "escapes | count | index | gindex | group |
+---------------+-------+-------+--------+-------+
| 0             | "A"   | 0     | 0      | 0     |
| 1             | "B"   | 1     | 1      | 0     |
| 2             | "C"   | 2     | 2      | 0     |
+---------------+-------+-------+--------+-------+
+-----+-----------+
| c0  | c1        |
+-----+-----------+
| "1" | (0,0,0,0) |
| "2" | (1,1,1,1) |
| "3" | (2,2,2)   |
| "4" | (3,3,3)   |
+-----+-----------+
+-----+-----+-----+
| a   | b   | c   |
+-----+-----+-----+
| "A" | "1" | "a" |
| "B" | "2" | "b" |
| "A" | "3" | "c" |
| "B" | "4" | "d" |
+-----+-----+-----+
+-----+-----+-----+
| a   | b   | c   |
+-----+-----+-----+
| "A" | "1" | "a" |
| "B" | "2" | "b" |
+-----+-----+-----+
+-----+-----+-----+
| a   | b   | c   |
+-----+-----+-----+
| "A" | "1" | "a" |
| "A" | "1" | "e" |
| "A" | "1" | "i" |
| "A" | "1" | "m" |
+-----+-----+-----+
(14,7)
+-----+-----+--------+
| a   | b   | gindex |
+-----+-----+--------+
| "A" | "1" | 0      |
| "A" | "3" | 1      |
| "A" | "1" | 2      |
| "A" | "3" | 3      |
| "A" | "1" | 4      |
| "A" | "3" | 5      |
| "A" | "1" | 6      |
| "B" | "2" | 7      |
| "B" | "4" | 8      |
| "B" | "2" | 9      |
| "B" | "4" | 10     |
| "B" | "2" | 11     |
| "B" | "4" | 12     |
| "B" | "2" | 13     |
+-----+-----+--------+
+-----+-----+--------+
| a   | b   | gindex |
+-----+-----+--------+
| "A" | "1" | 0      |
| "A" | "3" | 1      |
| "A" | "1" | 2      |
| "A" | "3" | 3      |
| "A" | "1" | 4      |
| "A" | "3" | 5      |
| "A" | "1" | 6      |
+-----+-----+--------+
+---+---+---+
| a | b | c |
+---+---+---+
+---+---+---+
| a | b | c |
+---+---+---+
+-------+
| value |
+-------+
+-------+
| value |
+-------+
| 42    |
+-------+
+----+-----+-----+
| a  | b   | c   |
+----+-----+-----+
| 11 | 22  | 33  |
| 44 | nil | nil |
+----+-----+-----+
+----+-------------+
| a  | quoted name |
+----+-------------+
| 55 | 66          |
| 77 | 88          |
+----+-------------+
+----+
| a  |
+----+
| 11 |
| 22 |
+----+
+-----------+
| __proto__ |
+-----------+
| 11        |
| 22        |
+-----------+
